File: python/src/datagrout/conduit/namespaces/warden.py
"""Warden namespace — safety checks, intent verification, and consensus."""

from typing import Any, Dict


class WardenNamespace:
    """Safety gates, intent verification, and multi-model consensus."""

    def __init__(self, client: Any):
        self._client = client

    async def canary(self, params: Dict[str, Any]) -> Any:
        """Run a canary safety check (`data-grout/warden.canary`)."""
        self._client._warn_if_not_dg("warden.canary")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/warden.canary", params)
        )

    async def verify_intent(self, params: Dict[str, Any]) -> Any:
        """Verify intent before executing an action (`data-grout/warden.intent`)."""
        self._client._warn_if_not_dg("warden.intent")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/warden.intent", params)
        )

    async def adjudicate(self, params: Dict[str, Any]) -> Any:
        """Adjudicate a dispute or ambiguity (`data-grout/warden.adjudicate`)."""
        self._client._warn_if_not_dg("warden.adjudicate")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/warden.adjudicate", params)
        )

    async def ensemble(self, params: Dict[str, Any]) -> Any:
        """Multi-model ensemble consensus check (`data-grout/warden.ensemble`)."""
        self._client._warn_if_not_dg("warden.ensemble")
        self._client._ensure_initialized()
        return await self._client._send_with_retry(
            lambda: self._client._transport.call_tool("data-grout/warden.ensemble", params)
        )

File: elixir/lib/datagrout_conduit/warden.ex
defmodule DatagroutConduit.Warden do
  @moduledoc "Safety gates, intent verification, and multi-model consensus."

  @doc "Run a canary safety check (`data-grout/warden.canary`)."
  @spec canary(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def canary(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "warden.canary", opts)
  end

  @doc "Verify intent before executing an action (`data-grout/warden.intent`)."
  @spec verify_intent(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def verify_intent(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "warden.intent", opts)
  end

  @doc "Adjudicate a dispute or ambiguity (`data-grout/warden.adjudicate`)."
  @spec adjudicate(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def adjudicate(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "warden.adjudicate", opts)
  end

  @doc "Multi-model ensemble consensus check (`data-grout/warden.ensemble`)."
  @spec ensemble(GenServer.server(), map()) :: {:ok, map()} | {:error, term()}
  def ensemble(client, opts \\ %{}) do
    DatagroutConduit.Client.dg(client, "warden.ensemble", opts)
  end
end
